wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevFileConverter.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.String.format;
+import static org.eclipse.che.api.devfile.server.Constants.CURRENT_SPEC_VERSION;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+import org.eclipse.che.api.devfile.model.Action;
+import org.eclipse.che.api.devfile.model.Command;
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.devfile.model.Project;
+import org.eclipse.che.api.devfile.model.Source;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.workspace.server.model.impl.CommandImpl;
+import org.eclipse.che.api.workspace.server.model.impl.ProjectConfigImpl;
+import org.eclipse.che.api.workspace.server.model.impl.SourceStorageImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+
+/** Helps to convert Devfile into workspace config and back. */
+public class DevFileConverter {
+
+  public Devfile workspaceToDevFile(WorkspaceConfigImpl wsConfig) {
+    Devfile devFile =
+        new Devfile().withSpecVersion(CURRENT_SPEC_VERSION).withName(wsConfig.getName());
+
+    // Manage projects
+    List<Project> projects = new ArrayList<>();
+    wsConfig
+        .getProjects()
+        .forEach(projectConfig -> projects.add(projectConfigToDevProject(projectConfig)));
+    devFile.setProjects(projects);
+
+    // Manage commands
+    List<Command> commands = new ArrayList<>();
+    wsConfig.getCommands().forEach(command -> commands.add(commandImplToDevCommand(command)));
+    devFile.setCommands(commands);
+
+    // Manage tools
+    List<Tool> tools = new ArrayList<>();
+    for (Map.Entry<String, String> entry : wsConfig.getAttributes().entrySet()) {
+      if (entry.getKey().equals("editor")) {
+        String editorId = entry.getValue();
+        Tool editorTool =
+            new Tool()
+                .withType("cheEditor")
+                .withId(editorId)
+                .withName(findToolName(wsConfig, editorId));
+        tools.add(editorTool);
+      } else if (entry.getKey().equals("plugins")) {
+        for (String pluginId : entry.getValue().split(",")) {
+          Tool pluginTool =
+              new Tool()
+                  .withId(pluginId)
+                  .withType("chePlugin")
+                  .withName(findToolName(wsConfig, pluginId));
```

Comment:
You parse tools aliases N times, where N is plugins number. Please consider parsing them once and save in Map.